igvm/puppet.py

```diff
@@ -62,9 +65,29 @@ def clean_cert(vm, user=None):
         version = sudo('/usr/bin/puppet --version', shell=False, quiet=True)
 
         if not version.succeeded or int(version.split('.')[0]) < 6:
-            sudo('/usr/bin/puppet cert clean {}'.format(
-                vm['hostname'],
-            ), shell=False)
+            res = sudo(
+                '/usr/bin/puppet cert verify {} '
+                '&& /usr/bin/puppet cert clean {}'.format(
+                    vm['hostname'], vm['hostname'],
+                ),
+            )
+
+            retry = 0
+            while res.return_code == 3 and retry < retries:
+                sleep(1)
+                retry += 1
+                logger.info(
+                    'Retrying to clear certificate (retry {}/{})'.format(
+                        retry, retries,
+                    )
+                )
+                res = run(
+                    'sudo /usr/bin/puppet cert verify {} '
+                    '&& sudo /usr/bin/puppet cert clean {}'.format(
+                        vm['hostname'], vm['hostname'],
+                    ),
+                    shell=False,
+                )
```

Comment:
How about:
```python
            for retry in range(0, retries):
                logger.info(
                    'Trying to clear certificate (retry {}/{})'.format(
                        retry+1, retries,
                    )
                )
                res = run(
                    'sudo /usr/bin/puppet cert verify {} '
                    '&& sudo /usr/bin/puppet cert clean {}'.format(
                        vm['hostname'], vm['hostname'],
                    ),
                    shell=False,
                )
                if res.return_code != 3:
                    break
```
This would remove the duplication of the puppet command.

Reply:
thank you :). i have to admit my solution is very ugly. yours is much better. will add it quickly